libweston/backend-rdp/rdputil.c

```diff
@@ -364,38 +364,126 @@ dump_id_manager_state(FILE *fp, struct rdp_id_manager *id_manager, char* title)
-		struct wl_event_loop *loop = wl_display_get_event_loop(b->compositor->wl_display);
-		*event_source = wl_event_loop_add_fd(loop,
-					peerCtx->loop_event_source_fd,
-					WL_EVENT_READABLE,
-					func, data);
-		if (*event_source) {
-			eventfd_write(peerCtx->loop_event_source_fd, 1);
-			return true;
-		} else {
-			rdp_debug_error(b, "%s: wl_event_loop_add_idle failed\n", __func__);
-		}
-	} else {
-		/* RDP server is not opened, this must not be used */
-		assert(false);
-	}
-	return false;
+	*event_source = wl_event_loop_add_fd(loop, fd, 0, func, data);
+	if (!*event_source)
+		return false;
+
+	wl_event_source_fd_update(*event_source, mask);
+	return true;
 }
 
 void
-rdp_defer_rdp_task_done(RdpPeerContext *peerCtx)
+rdp_dispatch_task_to_display_loop(RdpPeerContext *peerCtx, wl_event_loop_fd_func_t func, struct rdp_loop_task *task)
+{
+	/* this function is ONLY used to queue the task from FreeRDP thread,
+	   and the task to be processed at wayland display loop thread. */
+	ASSERT_NOT_COMPOSITOR_THREAD(peerCtx->rdpBackend);
+
+	task->peerCtx = peerCtx;
+	task->func = func;
+
+	pthread_mutex_lock(&peerCtx->loop_task_list_mutex);
+	/* this inserts at head */
+	wl_list_insert(&peerCtx->loop_task_list, &task->link);
+	pthread_mutex_unlock(&peerCtx->loop_task_list_mutex);
+
+	eventfd_write(peerCtx->loop_task_event_source_fd, 1);
+}
+
+static int
+rdp_dispatch_task(int fd, uint32_t mask, void *arg)
 {
+	RdpPeerContext *peerCtx = (RdpPeerContext *)arg;
+	struct rdp_loop_task *task, *tmp;
 	eventfd_t dummy;
-	eventfd_read(peerCtx->loop_event_source_fd, &dummy);
+
+	/* this must be called back at wayland display loop thread */
+	ASSERT_COMPOSITOR_THREAD(peerCtx->rdpBackend);
+
+	eventfd_read(peerCtx->loop_task_event_source_fd, &dummy);
+
+	pthread_mutex_lock(&peerCtx->loop_task_list_mutex);
+	/* dequeue the first task which is at last, so use reverse. */
+	assert(!wl_list_empty(&peerCtx->loop_task_list));
+	wl_list_for_each_reverse_safe(task, tmp, &peerCtx->loop_task_list, link) {
+		wl_list_remove(&task->link);
+		break;
+	}
+	pthread_mutex_unlock(&peerCtx->loop_task_list_mutex);
+
+	/* Dispatch and task will be freed by caller. */
+	assert(fd >= 0);
+	return task->func(fd, mask, task);
+}
+
+bool
+rdp_initialize_dispatch_task_event_source(RdpPeerContext *peerCtx)
+{
+	struct rdp_backend *b = peerCtx->rdpBackend;
+	struct wl_event_loop *loop;
+
+	assert(peerCtx->loop_task_list_mutex_initialized == false);
+	if (pthread_mutex_init(&peerCtx->loop_task_list_mutex, NULL) == -1) {
+		rdp_debug_error(b, "%s: pthread_mutex_init failed. %s\n", __func__, strerror(errno));
+		return false;
+	}
+	peerCtx->loop_task_list_mutex_initialized = true;
+
+	assert(peerCtx->loop_task_event_source_fd == -1);
+	peerCtx->loop_task_event_source_fd = eventfd(0, EFD_SEMAPHORE | EFD_CLOEXEC);
+	if (peerCtx->loop_task_event_source_fd == -1) {
+		rdp_debug_error(b, "%s: eventfd(EFD_SEMAPHORE) failed. %s\n", __func__, strerror(errno));
+		return false;

```

Comment:
As mentioned earlier in this review, I think if this function cleaned up internally instead of exiting with indeterminate state there would be some benefits.

There wouldn't be any need for the mutex_initialized_bool, and the following teardown function could always assume it was dealing with completely initialized state.

Reply:
Done